### PR TITLE
847 Investigation of connection sharing

### PR DESCRIPTION
**Pull request by whpearson**
We noticed lots of connections to the database and that we were creating
a new client each time we did a query.

This is an investigation into maintaining a persistent http connection
to aws.

It uses the flask g object to store a connection which has a the headers
set to keep alive.

### CODE REVIEW

**Review by sanisc — COMMENTED** [review]
what happens when the app is torn down does the connection still exist ?

**Comment by whpearson** [comment]
Ah, good point. I'll add some cleanup.

File: vulnerable_people_form/integrations/persistence.py [hunk 1]
```diff
@@ -22,8 +23,16 @@ def get_client_kwargs(app=current_app):
     }
 
 
+def set_connection_header(request, operation_name, **kwargs):
+    request.headers['Connection'] = 'keep-alive'
+
+
 def get_rds_data_client(app=current_app):
-    return boto3.client("rds-data", **get_client_kwargs(app))
+    if 'boto_rds_client' not in g:
```
Comment (VivianAllen) [inline vulnerable_people_form/integrations/persistence.py:31]: this is really neat! I presume its more reliable than just instantiating a global client on import?

Comment (whpearson) [reply vulnerable_people_form/integrations/persistence.py:31]: Just going by documentation on how to maintain db connections. https://flask.palletsprojects.com/en/1.1.x/appcontext/

You could have some retry logic here if the client gets in a bad state.

Comment (VivianAllen) [reply vulnerable_people_form/integrations/persistence.py:31]: cool stuff!

**Review by VivianAllen — APPROVED** [review]
lgtm ✅

**Comment by whpearson** [comment]
Going to hold off merging, until we have a run at the current level.

File: vulnerable_people_form/integrations/persistence.py (hunk 1 above)
Comment (sanisc) [reply vulnerable_people_form/integrations/persistence.py:31]: what happens if it does get a bad state and the app is still running ? Is that something we can pre-empt?

**Comment by whpearson** [comment]
Closing for now, as performance seems okay for our needs.

**Pull request closed without merging**